File: go.sdk/src/bma/servicecall/core/address.go
package core

import (
	"bytes"
	"fmt"
	"strings"
)

type Address struct {
	typ    string
	api    string
	option *ValueMap
}

func NewAddress() *Address {
	o := new(Address)
	return o
}

func CreateAddress(typ string, api string, opts map[string]interface{}) *Address {
	o := NewAddress()
	o.typ = typ
	o.api = api
	o.option = CreateValueMap(opts)
	return o
}

func CreateAddressFromMap(vm map[string]interface{}) *Address {
	return CreateAddressFromValue(CreateValueMap(vm))
}

func CreateAddressFromValue(vm *ValueMap) *Address {
	o := NewAddress()
	o.typ = vm.GetString("Type")
	o.api = vm.GetString("API")
	o.option = vm.GetMap("Option")
	return o
}

func (this *Address) Valid() error {
	if this.typ == "" {
		return fmt.Errorf("address type empty")
	}
	if this.api == "" {
		return fmt.Errorf("address api empty")
	}
	return nil
}

func (this *Address) GetType() string {
	return this.typ
}

func (this *Address) SetType(v string) {
	this.typ = v
}

func (this *Address) GetAPI() string {
	return this.api
}

func (this *Address) SetAPI(v string) {
	this.api = v
}

func (this *Address) GetOption() *ValueMap {
	return this.option
}

func (this *Address) SetOption(v *ValueMap) {
	this.option = v
}

func (this *Address) SureOption() *ValueMap {
	if this.option == nil {
		this.option = NewValueMap(nil)
	}
	return this.option
}

func (this *Address) String() string {
	buf := bytes.NewBuffer([]byte{})
	buf.WriteString(fmt.Sprintf("TYPE:%s, API:%s", this.typ, this.api))
	if this.option != nil {
		buf.WriteString(fmt.Sprintf(", %s", this.option.Dump()))
	}
	return buf.String()
}

func (this *Address) ToValueMap() *ValueMap {
	vm := NewValueMap(nil)
	vm.Put("Type", this.typ)
	vm.Put("API", this.api)
	if this.option != nil && this.option.Len() > 0 {
		vm.Put("Option", this.option)
	}
	return vm
}

// Builder
type SimpleAddressBuilder struct {
	Type   string
	API    string
	Option *ValueMap
}

func NewAddressBuilder() *SimpleAddressBuilder {
	o := new(SimpleAddressBuilder)
	return o
}

func (this *SimpleAddressBuilder) Build(service, method string) *Address {
	s := strings.Replace(this.API, "$SNAME$", service, -1)
	s = strings.Replace(s, "$MNAME$", method, -1)
	r := NewAddress()
	r.SetType(this.Type)
	r.SetAPI(s)
	r.SetOption(this.Option)
	return r
}

File: java.sdk/core/src/main/java/bma/servicecall/core/Address.java
package bma.servicecall.core;

import java.util.Map;

public class Address {
	private String type;
	private String api;
	ValueMap option;

	public Address() {
		super();
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public String getApi() {
		return api;
	}

	public void setApi(String api) {
		this.api = api;
	}

	public ValueMap getOption() {
		return option;
	}

	public void setOption(ValueMap option) {
		this.option = option;
	}

	public ValueMap sureOption() {
		if (this.option == null) {
			this.option = new ValueMap(null);
		}
		return this.option;
	}

	@SuppressWarnings("rawtypes")
	public static Address createAddress(String type, String api, Map opts) {
		Address o = new Address();
		o.type = type;
		o.api = api;
		o.option = ValueMap.create(opts);
		return o;
	}

	@SuppressWarnings("rawtypes")
	public static Address createAddressFromMap(Map vm) {
		return createAddressFromValue(ValueMap.create(vm));
	}

	public static Address createAddressFromValue(ValueMap vm) {
		Address o = new Address();
		o.type = vm.getString("Type");
		o.api = vm.getString("API");
		o.option = vm.getMap("Option");
		return o;
	}

	public void Valid() {
		if (this.type == "") {
			throw new AppError("address type empty");
		}
		if (this.api == "") {
			throw new AppError("address api empty");
		}
		return;
	}

	public ValueMap toValueMap() {
		ValueMap vm = new ValueMap(null);
		vm.put("Type", this.type);
		vm.put("API", this.api);
		if (this.option != null && this.option.len() > 0) {
			vm.put("Option", this.option);
		}
		return vm;
	}

	@Override
	public String toString() {
		StringBuffer buf = new StringBuffer();
		buf.append("Address(");
		buf.append("Type:").append(this.type).append(";");
		buf.append("API:").append(this.api).append(";");
		buf.append(")");
		return buf.toString();
	}
}
